src/smt/optimization_solver.cpp

```diff
@@ -0,0 +1,137 @@
+ * default objective constructed with Null Node and OBJECTIVE_UNDEFINED
+ *
+ * d_savedValue is initialized to a default node (Null Node)
+ */
+
+OptimizationSolver::OptimizationSolver(SmtEngine* parent)
+    : d_parent(parent),
+      d_activatedObjective(Node(), OBJECTIVE_UNDEFINED),
+      d_savedValue()
+{
+}
+
+OptimizationSolver::~OptimizationSolver() {}
+
+OptResult OptimizationSolver::checkOpt()
+{
+  // Make sure that the objective is not the default one
+  Assert(d_activatedObjective.getType() != OBJECTIVE_UNDEFINED);
+  Assert(!d_activatedObjective.getNode().isNull());
+
+  // the smt engine to which we send intermediate queries
+  // for the linear search.
+  std::unique_ptr<SmtEngine> optChecker;
+  initializeSubsolver(optChecker);
+  NodeManager* nm = optChecker->getNodeManager();
+
+  // we need to be in incremental mode for multiple objectives since we need to
+  // push pop we need to produce models to inrement on our objective
+  optChecker->setOption("incremental", "true");
+  optChecker->setOption("produce-models", "true");
+
+  // Move assertions from the parent solver to the subsolver
+  std::vector<Node> axioms = d_parent->getExpandedAssertions();
+  for (const Node& e : axioms)
+  {
+    optChecker->assertFormula(e);
+  }
+
+  // We need to checksat once before the optimization loop so we have a
+  // baseline value to increment
+  Result loop_r = optChecker->checkSat();
+
+  if (loop_r.isUnknown())
+  {
+    return OPT_UNKNOWN;
+  }
+  if (!loop_r.isSat())
+  {
+    return OPT_UNSAT;
+  }
+
+  // Model-value of objective (used in optimization loop)
+  Node value;
+  // asserts objective > old_value (used in optimization loop)
+  Node increment;
+
+  // Workhorse of linear optimization:
+  // This loop will keep incrmenting the objective until unsat
+  // When unsat is hit, the optimized value is the model value just before the
+  // unsat call
+  while (loop_r.isSat())
+  {
+    // get the model-value of objective in last sat call
+    value = optChecker->getValue(d_activatedObjective.getNode());
+
+    // We need to save the value since we need the model value just before the
+    // unsat call
+    Assert(!value.isNull());
+    d_savedValue = value;
+
+    // increment on the model-value of objective:
+    // if we're maximizing increment = objective > old_objective value
+    // if we're minimizing increment = objective < old_objective value
+    if (d_activatedObjective.getType() == OBJECTIVE_MAXIMIZE)
+    {
+      increment = nm->mkNode(kind::GT, d_activatedObjective.getNode(), value);
+    }
+    else
+    {
+      increment = nm->mkNode(kind::LT, d_activatedObjective.getNode(), value);
+    }
+    optChecker->assertFormula(increment);
+    loop_r = optChecker->checkSat();
+  }
+
+  return OPT_OPTIMAL;
+}
+
+void OptimizationSolver::activateObj(const Node& obj, const int& type)
+{
+  d_activatedObjective = Objective(obj, (ObjectiveType)type);
+}
+
+Node OptimizationSolver::objectiveGetValue(const Node& obj)
```

Comment:
Does this method need to take obj?  It is not used here.  If this method took no arguments, then we can just assume you are looking up the value from the last query.